logic/dungeon4.py

```diff
@@ -10,29 +10,41 @@ def __init__(self, options):
         entrance.add(DungeonChest(0x16A))  # map chest
         right_of_entrance = Location(4).add(DungeonChest(0x178)).connect(entrance, AND(SHIELD, attack_hookshot)) # 2 zol 1 spike enemy
         Location(4).add(DungeonChest(0x17B)).connect(right_of_entrance, SWORD) # room with key chest
-        Location(4).add(DungeonChest(0x171)).add(DungeonChest(0x165)).connect(entrance, AND(FEATHER, PEGASUS_BOOTS, BOMB))  # 2 key chests on the right.
-
+        rightside_crossroads = Location(4).connect(entrance, AND(FEATHER, PEGASUS_BOOTS))  # 2 key chests on the right.
+        pushable_block_chest = Location(4).add(DungeonChest(0x171)).connect(rightside_crossroads, BOMB) # lower chest
+        puddle_crack_block_chest = Location(4).add(DungeonChest(0x165)).connect(rightside_crossroads, OR(BOMB, FLIPPERS)) # top right chest
         double_locked_room = Location(4).connect(right_of_entrance, KEY4)
         after_double_lock = Location(4).connect(double_locked_room, AND(KEY4, OR(FEATHER, FLIPPERS)))
-        Location(4).add(DungeonChest(0x175)).connect(after_double_lock, FLIPPERS)
-        before_miniboss = Location(4).connect(after_double_lock, AND(FEATHER, PEGASUS_BOOTS, KEY4))
+        dungeon4_puddle_before_crossroads = Location(4).add(DungeonChest(0x175)).connect(after_double_lock, FLIPPERS)
+        north_crossroads = Location(4).connect(after_double_lock, AND(FEATHER, PEGASUS_BOOTS))
+        before_miniboss = Location(4).connect(north_crossroads, AND(KEY4, FOUND(KEY4, 3)))
         Location(4).add(OwlStatue(0x16F)).connect(before_miniboss, STONE_BEAK4)
         sidescroller_key = Location(4).add(DroppedKey(0x169)).connect(before_miniboss, FLIPPERS)  # key that drops in the hole and needs swim to get
         Location(4).add(DungeonChest(0x16E)).connect(before_miniboss, FLIPPERS)  # chest with 50 rupees
         before_miniboss.add(DungeonChest(0x16D))  # gel chest
         before_miniboss.add(DungeonChest(0x168))  # key chest near the puzzle
-        miniboss = Location(4).connect(before_miniboss, AND(KEY4, POWER_BRACELET))
+        miniboss = Location(4).connect(before_miniboss, OR(FLIPPERS, AND(KEY4, POWER_BRACELET))) # move around miniboss through 5 tile room
         miniboss.add(DungeonChest(0x160))  # flippers chest
 
-        to_the_nightmare_key = Location(4).connect(before_miniboss, FLIPPERS)  # 5 symbol puzzle
+        to_the_nightmare_key = Location(4).connect(before_miniboss, AND(FEATHER, OR(FLIPPERS, PEGASUS_BOOTS)))  # 5 symbol puzzle (does not need flippers with boots + feather)

```

Comment:
I would say it's ok, It is a 3 gap jump from the bottom to the top left switch, the others are freely accessibly. you do have to jump from like the top half of the tile to make it (because water instead of pits), but I do not believe its extreme to expect to jump 3 tiles straight over water with boots and feather